Handle SNS notifications for emails sent

albertisfu opened the pull request:
This PR contains:
- Signals to handle:
 -Complaints notifications
 -Bounces notifications
 -Delivery notifications

- Tests to verify notification events are received and handled properly according to each type of event.

- Function to handle soft bounces notifications
-`EmailFlag` and `BackoffEvent` models were added:
  During tests, I realized that storing user_id in these models are not necessary because we are going to flag and ban emails, not users, and there could exist some conflict cases, for example when a user changes its email and then we receive a bounce or complaint related with the previous user email address we won't be able to assign this event to a user because the user's email address changed. 
  I think it was better to create a date_created field directly in the model instead of using  `AbstractDateTimeModel` because we are not going to use a `date_modified` field and also I think we don't need  index for these datetime fields.

- Tests to verify soft bounces function works properly under each type of soft bounce notifications and different scenarios.

albertisfu commented:
@mlissner added a commit in this PR for phase 2: handle hard bounces and delivery notifications because it also contains some improvements for previous commits.

I noticed I didn't explain how exponential backoff is working:

- The general idea is to have an average max retry period of 5 days the exact period will depend on how fast SES sends bounce notifications.

- The first time we get a soft bounce notification we are going to wait for 2 hours before trying to send another email to the same email address.

- If after sending a new email we get another soft bounce notification for the same email address the next waiting period will be 4 hours, if continue bouncing the next waiting period will be 8, 16, 32, and 64 hours in the 5º retry, for a total of 126 hours (in average 5.25 days) since the initial bounce, after the 5º retry if we receive another bounce notification the email address is banned.

Noticed there are some conflicts to merge with main, I'll check them and solve them.

mlissner commented on cl/users/models.py:
```diff
@@ -190,8 +259,211 @@ def generate_recap_email(user_profile: UserProfile, append: int = None) -> str:
+            )
+            # After log the event ban the email address
+            # Only ban email address if it hasn't been previously banned
+            EmailFlag.objects.get_or_create(
+                email_address=email,
+                flag_type="ban",
+                defaults={"reason": bounceSubType},
+            )
+        else:
+            # Any other hard bounce even only ban the email address
+            # Only ban email address if it hasn't been previously banned
+            EmailFlag.objects.get_or_create(
+                email_address=email,
+                flag_type="ban",
+                defaults={"reason": bounceSubType},
+            )
```
You can just remove the `else` statement and outdent the `get_or_create` and this will be more clear. Also, please use `flag_type=EmailFlag.BAN` instead of spelling out the choice.

albertisfu replied:
Fixed! btw I moved all these email logic code to cl.users.email_handlers and cl.users.signals.py

mlissner commented on cl/users/models.py:
```diff
@@ -190,8 +259,211 @@ def generate_recap_email(user_profile: UserProfile, append: int = None) -> str:
+def handle_hard_bounce(
+    message_id: str, bounceSubType: str, recipient_emails: List[str]
+) -> None:
+    """Handle a hard bounce notification received from SNS
```
```suggestion
    """Ban any email address that receives a hard bounce.
    
    If the bounce is of an unexpected type, also log a warning.
    
```

mlissner commented on cl/users/models.py:
```diff
@@ -190,8 +259,211 @@ def generate_recap_email(user_profile: UserProfile, append: int = None) -> str:
+) -> None:
+    """Handle a hard bounce notification received from SNS
+    :param message_id: The unique message id assigned by Amazon SES
+    :param bounceSubType: The subtype of the bounce, as determined
```
I think I'd convert this to snake_case instead of camelCase even though AWS prefers to camelCase.

albertisfu replied:
Done!

mlissner commented on cl/users/models.py:
```diff
@@ -190,8 +259,211 @@ def generate_recap_email(user_profile: UserProfile, append: int = None) -> str:
+def handle_soft_bounce(
+    message_id: str, bounceSubType: str, recipient_emails: List[str]
+) -> None:
+    """Handle a soft bounce notification received from SNS
```
Should have a newline here:
```suggestion
    """Handle a soft bounce notification received from SNS
    
```

mlissner commented on cl/users/models.py:
```diff
@@ -190,8 +259,211 @@ def generate_recap_email(user_profile: UserProfile, append: int = None) -> str:
+) -> None:
+    """Handle a soft bounce notification received from SNS
+    :param message_id: The unique message id assigned by Amazon SES
+    :param bounceSubType: The subtype of the bounce, as determined
```
Again, snake_case, not camelCase, please.

albertisfu replied:
Done!

mlissner commented on cl/users/models.py:
```diff
@@ -190,8 +259,211 @@ def generate_recap_email(user_profile: UserProfile, append: int = None) -> str:
+    :return: None
+    """
+
+    back_off_events = ["Undetermined", "General", "MailboxFull"]
```
Adding a comment referencing this issue or the other place we discussed this would be helpful here to help explain the different types and what we do. A long comment here might not hurt.

https://github.com/freelawproject/courtlistener/issues/1942

albertisfu replied:
Done! I added a detailed comment explaining how this works.

mlissner commented on cl/users/models.py:
```diff
@@ -190,8 +259,211 @@ def generate_recap_email(user_profile: UserProfile, append: int = None) -> str:
+                next_retry_date = backoff_event[0].next_retry_date
+
+                # Check if waiting period expired
+                if next_retry_date <= now():
```
I think flipping this is a bit more clear, so it reads as:

> If the time now is after the retry date

Instead of:

> If the retry date is before now

Same meaning, but the first one makes more sense to me. So:

```suggestion
                if now() >= next_retry_date:
```

It *might* be even more clear to do:

```
if now() < next_retry_date:
    # It's not yet time to retry. 
    continue
```

Then you can outdent all that follows. 

I'm also curious, are we missing the `else` statement here? What if it's not yet time to send the email? Do we save it or something?

albertisfu replied:
oh yes it's more clear like you suggested `if now() >= next_retry_date:` 

I used this option instead of `if now() < next_retry_date:` because here we need to act when the waiting period is over,  so we don't have an `else`, here is a screenshot where I show this part of the code, the `else` is "Don't update the backoff event" so we are just ignoring it, as you could see we have a `Queue Email` function before entering to these `if `statements, so yes we are going to queue this email to send it later.
 I put a comment in the code where we are going to call this function `#TODO Queue email function`

![screen_1](https://user-images.githubusercontent.com/486004/159105942-70d122ec-56cc-4ad8-9a43-6c9fda14d3de.png)

mlissner commented on cl/users/models.py:
```diff
@@ -190,8 +259,211 @@ def generate_recap_email(user_profile: UserProfile, append: int = None) -> str:
+                        new_next_retry_date = now() + timedelta(
+                            hours=pow(INITIAL_HOURS, new_retry_counter + 1)
+                        )
+                        BackoffEvent.objects.filter(
+                            email_address=email,
+                        ).update(
+                            retry_counter=new_retry_counter,
+                            next_retry_date=new_next_retry_date,
+                        )
```
There's a race condition here, but I suppose it's fine. Easier to live with it than to fix it, probably.

albertisfu replied:
I see, I definitely would like to fix this, would it be helpful to use `select_for_update()` and `@transaction.atomic`?
I would be very grateful if you could give me a hint as to whether this is a good approach or if there might be a better one.

mlissner replied:
Yes, I think `select_for_update` would be a simple and good way to do this.

mlissner commented on cl/users/tests.py:
```diff
@@ -329,3 +334,498 @@ def test_set_password_using_the_HTML(self) -> None:
+
+class SNSWebhookTest(TestCase):
+    def setUp(self) -> None:
+        test_dir = os.path.join(
```
I'm constantly encouraging people to use Pathlib for things like this, but dont' feel obligated to upgrade this code. One day I'll get mean about it, I guess. :)

albertisfu replied:
Solved! I'm now using Pathlib :)

mlissner reviewed (changes requested):
OK! Lots of little stuff, but nothing too big. In general, it looks great and I'm excited for the next stage to land. 

A few high-level items though:

1. If you can add admin files for the email models, that'd be great. 

2. Let's move all the email logic out of the models.py file. Maybe put it in cl.users.email_handlers.py?

3. We need a rebase, unfortunately, but hopefully that won't be too bad.

Generally looks good! Thank you, Alberto!

albertisfu commented on cl/users/urls.py:
```diff
@@ -137,4 +138,9 @@
+    path(
+        "ses/event-webhook/",
+        SESEventWebhookView.as_view(),
+        name="handle-event-webhook",
```
I agree, solved!

albertisfu commented on docker/django/Dockerfile:
```diff
@@ -54,7 +54,7 @@ RUN apt-get update --option "Acquire::Retries=3" --quiet=2 && \
         imagemagick tesseract-ocr tesseract-ocr-eng \
         # Other dependencies
-        libffi-dev libxml2-dev libxslt-dev procps vim cmake
+        libffi-dev libxml2-dev libxslt-dev procps vim cmake libssl-dev swig
```
Done, I put it like this: 
`# For installing M2Crypto necessary for django-ses`
` libssl-dev swig`

albertisfu commented on cl/users/models.py:
```diff
@@ -174,6 +180,69 @@ class Meta:
+        help_text="The notification subtype",
+        max_length=50,
+        blank=True,
+    )
```
Done! I added a class `SUB_TYPES` to handle these choices, also changed the field name from `reason` to `event_sub_type` because we are going to use for bounces and complaints subtypes

albertisfu commented on cl/users/models.py:
```diff
@@ -174,6 +180,69 @@ class Meta:
+    date_created = models.DateTimeField(
+        help_text="The moment when the item was created.",
+        auto_now_add=True,
+    )
```
Thank you for the detailed explanation of why we need to use `AbstractDateTimeModel`.
I have updated the model to inherit from `AbstractDateTimeModel` and remove the old `date_created` field.

albertisfu commented on cl/users/models.py:
```diff
@@ -174,6 +180,69 @@ class Meta:
 
 
+class EmailFlag(models.Model):
+    BAN = "ban"
```
Done, I added a detailed comment.

albertisfu commented on cl/users/models.py:
```diff
@@ -174,6 +180,69 @@ class Meta:
+        (BAN, "Email ban"),
+        (FLAG, "Email flag"),
+    )
+    email_address = models.EmailField(
+        help_text="Email address flagged.",
```
Done, I added a detailed comment, like this:
```
 :email_address: an EmailFlag object belongs to an email address instead
    of a user, in this way, if users change their email address this won't
    affect new user email address.
```

albertisfu commented on cl/users/models.py:
```diff
@@ -174,6 +180,69 @@ class Meta:
+        help_text="The flag type assigned.",
+        choices=FLAGS_TYPES,
+        max_length=5,
+    )
```
Fixed, I use a SmallIntegerField for these choice fields.

albertisfu commented on cl/users/models.py:
```diff
@@ -174,6 +180,69 @@ class Meta:
+        help_text="The flag assigned to email address.",
+        max_length=25,
+        blank=True,
+    )
```
Done, I added a comment like this: `The actual flag assigned, like: small_email_only` and also changed the field to be also an integer choice field, for now, we just use two flags "small_email_only" and "max_retry_reached"

albertisfu commented on cl/users/models.py:
```diff
@@ -174,6 +180,69 @@ class Meta:
+
+
+class BackoffEvent(models.Model):
+    email_address = models.EmailField(
```
Done, I added the comment.

albertisfu commented:
Thank you so much for your feedback Mike, I really appreciate it!

I think most of the changes are now solved, I left some comments explaining some of them, it is pending the one related to the race condition, also left a question about it.
Also solved:
- Moved email logic and signals outside of the models to cl.users.email_handlers and cl.users.signals.py
- Rebased with main, (poetry.lock) conflicts solved

mlissner commented on cl/users/models.py:
```diff
@@ -174,6 +205,86 @@ class Meta:
+class EmailFlag(AbstractDateTimeModel):
+    """Stores flags for email addresses
+
+    Two types of :object_type:
+    BAN: ban an email address and avoid sending any email
+    FLAG: flag an email address for a special treatment e.g small_email_only.
+    :flag: the actual flag assigned, like: small_email_only
+    :event_sub_type: the SNS bounce subtype that triggered the object
+    :email_address: an EmailFlag object belongs to an email address instead
+    of a user, in this way, if users change their email address this won't
+    affect new user email address.
```
For the field level details, any reason to put it here instead of in the help_text or in a comment near the field? I think it moves the comment further from where you need it. It's also not how we usually have done it, but if this is a pattern elsewhere, I'd be happy to consider a different approach.

mlissner commented on cl/users/models.py:
```diff
@@ -174,6 +205,86 @@ class Meta:
+    """Stores backoff events for email addresses, this is created or updated
+    after receiving a soft bounce object for an email address.
+
+    :email_address: The backoff event is related to this email address
+    instead of a user, in this way, if users change their email address
+    this won't affect new user email address.
```
I'd suggest moving this too, unless this is a common style I just haven't seen.

mlissner commented:
Great, thank you. Just a couple more comments at the moment. Once you've got those in I'll do one more full review and hopefully we'll be good. I also wanted to share this with you:

https://github.com/django-ses/django-ses/issues/182

I think it'd be pretty easy to do, but I'm waiting for the maintainer to give us the OK. Once done, we can remove the docker tweaks, but I don't know when they'll reply, so I think we have to move forward with m2crypto for now.

albertisfu commented:
I've solved the last two comments.

Wonderful I saw that you have already started working with `django-ses` contribution, if you need me to check something about it, let me know.

mlissner commented on cl/users/models.py:
```diff
@@ -22,6 +23,36 @@
 ]
 
 
+class SUB_TYPES(object):
+    """SNS Event Subtypes"""
+
+    UNDETERMINED = 0
+    GENERAL = 1
+    NOEMAIL = 2
+    SUPPRESSED = 3
+    ONACCOUNTSUPPRESSIONLIST = 4
+    MAILBOXFULL = 5
+    MESSAGETOOLARGE = 6
+    CONTENTREJECTED = 7
+    ATTACHMENTREJECTED = 8
+    COMPLAINT = 9
+    OTHER = 10
+
+    TYPES = (
+        (UNDETERMINED, "Undetermined"),
+        (GENERAL, "General"),
+        (NOEMAIL, "NoEmail"),
+        (SUPPRESSED, "Suppressed"),
+        (ONACCOUNTSUPPRESSIONLIST, "OnAccountSuppressionList"),
+        (MAILBOXFULL, "MailboxFull"),
+        (MESSAGETOOLARGE, "MessageTooLarge"),
+        (CONTENTREJECTED, "ContentRejected"),
+        (ATTACHMENTREJECTED, "AttachmentRejected"),
+        (COMPLAINT, "Complaint"),
+        (OTHER, "Other"),
+    )
```
This should be closer to the model it's used in, not at the top of the file.

mlissner commented:
I just caught one more little thing and I replied to the race condition thing. Otherwise, I think we're good without further review. Thank you!

albertisfu commented:
- I've solved the race condition thing using `selec_for_update()` and `transaction.atomic`.

- I also added some constraints to models to make them unique and avoid duplicates, for `BackoffEvent` the constraint is by email_address.

- For `EmailFlag` the constraint is by email and `object_type = BAN`, this way it'll only be possible to have one BAN object per email_address, added a test to ensure we can update a banned email instead of creating a new object.

- I've rebased with main and solved poetry conflicts because I'm going also to rebase this with the custom backend email branch that I'm working on to test them together.

Seems there's an issue with pre-commit.ci check, but I think it came from main

mlissner commented:
This looks ready to merge to me, but I think if we wait another 24 hours the new version of django-ses will remove the m2crypto junk, so I think we should wait for that, then fix the m2crypto import and then press ahead. Bill is also working on the docker image, so he might appreciate it not changing under him either, if we can help it. But no further tweaks on this from me.

albertisfu commented:
I've added support for multiple addresses for `to`, `bcc`, `cc` and `reply_to` fields.

Now we can store and handle banned, small emails and backoff events for multiple recipients.

- When storing messages we can store the original `from` email address in this format: 
   `Admin User <success@simulator.amazonses.com>` or just as a simple email `address success@simulator.amazonses.com`

- For `to`, `bcc`, `cc` and `reply to` fields, we normalize the email addresses before sending and storing the message, an email address like this `["Admin User <success@simulator.amazonses.com>"]` turns to `  ["success@simulator.amazonses.com"]`

When sending to multiple recipients we apply the following actions:

- For banned email: we verify if a recipient is banned, if so we remove the banned email addresses from the list of recipients to avoid sending emails to those addresses, if all of the recipients are banned, we discard the message, we don't send or store it.

- For small email only: if at least one of the recipients requires a small version, we send the small email version to all recipients.

- For email addresses under a backoff waiting period: if we detect an email address that is under a backoff waiting period we remove that address from the recipient list to avoid sending to it, we'll queue the message for those recipients.

mlissner reviewed (changes requested):
OK, this is ready except for *one* last thing. Can you please delete all your migrations and re-do them as a single one? Should be as simple as (roughly):

```
manage.py migrate users 0003
rm users/migrations/000{4..8}.*
manage.py migrate users -n add_email_handling
manage.py sqlmigrate users 0004
```

Migrations take time to run, so I try to avoid having more than one in a PR. They pile up!

mlissner commented:
Thank you @albertisfu! Huge PR, and a lot of work. Our email system is going to be incredible.
